# Lilsim Python SDK Test Suite

This notebook comprehensively tests all communication features of the lilsim simulator.

**Prerequisites:**
1. Start the lilsim application: `./build/debug/app/lilsim`
2. The tests will automatically enable/disable ZMQ and switch modes
3. Watch the GUI Status panel to see communication state updates

**What's Being Tested:**
- Connection and state streaming
- Admin commands (pause, run, reset, step)
- Asynchronous control mode
- Synchronous control mode
- Mode switching
- Client disconnection and timeout behavior
- Control period configuration (milliseconds → ticks)
- Marker visualization
- State updates continuity

In [1]:
# Setup and imports
import sys
import time
import numpy as np
import matplotlib.pyplot as plt
from IPython.display import clear_output

# Add the SDK to path
sys.path.insert(0, '.')

from lilsim import LilsimClient, AdminCommandType, MarkerType, ControlRequest
from lilsim.utils import state_to_dict

print("✓ Imports successful!")

✓ Imports successful!


/home/will/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.
  warnings.warn("Unable to import Axes3D. This may be due to multiple versions of "


In [2]:
client = LilsimClient(host="localhost")
client.connect()

INFO:lilsim.client:Connecting to lilsim at localhost...
INFO:lilsim.client:Connected to state stream (port 5556)
INFO:lilsim.client:Connected to admin command endpoint (port 5558)
INFO:lilsim.client:Connected to async control stream (port 5559)
INFO:lilsim.client:Connected to marker stream (port 5560)


In [3]:
client.reset()
client.run()

INFO:lilsim.client:Admin command succeeded: Reset requested
INFO:lilsim.client:Admin command succeeded: Simulation running


True

In [4]:
client.set_mode(sync=False)
client.send_control_async(steer_rate=-0.00, ax=-0)

INFO:lilsim.client:Admin command succeeded: Asynchronous mode enabled


In [5]:
# client.set_mode(sync=True, control_period_ms=100)
# def sinusoidal_controller(request: ControlRequest) -> tuple[float, float, float]:
#     tick = request.header.tick
#     sim_time = request.header.sim_time
    
#     # Sinusoidal steering
#     steer_rate = 0.3 * np.sin(tick * 0.02)
#     ax = 2.0
#     return (0.0, steer_rate, ax) # steer_angle, steer_rate, ax
# client.register_sync_controller(sinusoidal_controller)

# state_history = []
# def state_callback(state):
#     state_dict = state_to_dict(state)
#     state_history.append(state_dict)

# client.subscribe_state(state_callback)

# client.start()

In [6]:
# client.stop()

## Test: Visualization Markers

This test verifies marker publishing works correctly (port 5560).

In [17]:
client.clear_markers()

In [14]:
client.delete_namespace("test")

In [13]:
client.delete_marker("test", 1)

In [15]:
print("\nTesting visualization markers...")

# Circle marker
print("  - Publishing red circle at (5, 5)...")
client.publish_marker(
    ns="test", id=1, marker_type=MarkerType.CIRCLE,
    x=5.0, y=5.0, scale_x=2.0, scale_y=2.0,
    r=255, g=0, b=0, a=200
)

# Line strip forming a square
print("  - Publishing green square at (10, 10)...")
square = [(10,10,0), (12,10,0), (12,12,0), (10,12,0), (10,10,0)]
client.publish_marker(
    ns="test", id=2, marker_type=MarkerType.LINE_STRIP,
    points=square, r=0, g=255, b=0, a=255
)

# Circle at origin
print("  - Publishing blue circle at origin...")
client.publish_marker(
    ns="test", id=3, marker_type=MarkerType.CIRCLE,
    x=0.0, y=0.0, scale_x=1.5, scale_y=1.5,
    r=0, g=0, b=255, a=200
)

print("✓ Markers published - check viewport for red circle, green square, and blue circle")


Testing visualization markers...
  - Publishing red circle at (5, 5)...
  - Publishing green square at (10, 10)...
  - Publishing blue circle at origin...
✓ Markers published - check viewport for red circle, green square, and blue circle


In [20]:
# Line strip forming a square
print("  - Publishing green square at (10, 10)...")
square = [(10,10,0), (12,10,0), (12,12,0), (10,12,0), (10,10,0)]
client.publish_marker(
    ns="test", id=2, marker_type=MarkerType.LINE_STRIP,
    points=square, r=0, g=255, b=0, a=255
)


  - Publishing green square at (10, 10)...
